### PR TITLE
feat: return custom error

### PR DESCRIPTION
**Pull request by cullenmcdermott**
Hi! Thanks for this library. I'm currently using it in https://github.com/cullenmcdermott/terraform-provider-porkbun and someone reported they were [hitting rate limits](https://github.com/cullenmcdermott/terraform-provider-porkbun/issues/23) from Porkbun. 

I was having trouble parsing the error out so I've made some updates to return a custom error and to tell us if it should be retryable or not. I haven't added additional tests for this but I'm happy to if you'd like.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -21,6 +22,8 @@ const statusSuccess = "SUCCESS"
 // DefaultTTL The minimum and the default is 300 seconds.
 const DefaultTTL = "300"
 
+var retryableCodes = []int{429, 502, 503, 504}
```
Comment (ldez) [inline client.go]: The 50x statuses are not retryable by definition: they are server errors.

File: client.go [hunk 2]
```diff
@@ -209,8 +212,16 @@ func (c *Client) do(ctx context.Context, endpoint *url.URL, apiRequest interface
 	}
 
 	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("%d: %s", resp.StatusCode, string(respBody))
+		return nil, StatusError{
+			Status:    resp.StatusCode,
+			Message:   string(respBody),
+			Retryable: isRetryable(resp.StatusCode),
```
Comment (ldez) [inline client.go]: I think the status is enough, to decide if you want to retry or not, so the field `Retryable` seems useless.

**Review by ldez — CHANGES_REQUESTED** [review]

**Comment by ldez** [comment]
In summary, I'm ok with a typed error but the "retryable" part should be removed.

File: client.go (hunk 1 above)
Comment (cullenmcdermott) [reply client.go]: In certain circumstances retrying server errors make sense. In this situation Porkbun returns a 503 when you're being rate limited so we want to retry it. But I'm good making the changes, will push a commit shortly.

File: client.go (hunk 2 above)
Comment (cullenmcdermott) [reply client.go]: Fixed it.

**Comment by ldez** [comment]
Can you give me an example of the response body related to the rate limit?
Can you add a test?

**Comment by cullenmcdermott** [comment]
Running this command:
```
while true; do curl --fail -i --data @secret.json https://porkbun.com/api/json/v3/ping; done
```

secret.json just contains my apikey and secretapikey.

Returns this on the 3rd or 4th loop
```
HTTP/2 503 
date: Sun, 08 Jan 2023 00:04:24 GMT
content-type: text/html
content-length: 194
set-cookie: AWSALB=foobar; Expires=Sun, 15 Jan 2023 00:04:24 GMT; Path=/
set-cookie: AWSALBCORS=foobar; Expires=Sun, 15 Jan 2023 00:04:24 GMT; Path=/; SameSite=None; Secure
server: openresty
strict-transport-security: max-age=63072000; includeSubDomains; preload
x-frame-options: deny

curl: (22) The requested URL returned error: 503
```

I'm going to work on a test at some point this weekend hopefully.

File: client_test.go [hunk 3]
```diff
@@ -173,3 +173,11 @@ func TestClient_RetrieveRecords_error(t *testing.T) {
...
 	require.Error(t, err)
 }
+
+func TestClient_ErrorStatus(t *testing.T) {
+	client := setup(t, "/dns/retrieve/example.com", "error503")
+	_, err := client.RetrieveRecords(context.Background(), "example.com")
+	assert.IsType(t, &StatusError{}, err)
+	statusE := err.(*StatusError)
+	assert.Equal(t, "503", statusE.Status)
+}
```
Comment (ldez) [inline client_test.go]: based on your curl, this test is not realistic.

**Review by ldez — CHANGES_REQUESTED** [review]

File: client_test.go (hunk 3 above)
Comment (cullenmcdermott) [reply client_test.go]: Could you please provide some guidance on this then? I'm not sure what is unrealistic or how I would improve it.

Comment (cullenmcdermott) [reply client_test.go]: Hi! Just wanted to bump this to see if you could provide some guidance about what is unrealistic? Thanks!

**Comment by cullenmcdermott** [comment]
Also I just made some updates to fix warnings and some errors in CI. Happy to revert/change those if you prefer! This made everything pass on my fork.

**Review by ldez — APPROVED** [review]
LGTM

**Pull request merged**